### PR TITLE
CV hooks

### PR DESCRIPTION
**Pull request by jduerholt**
This PR adds cross validation hooks. By this it is possible to call callables from inside the CV loop. This can be used for example to compute cross validated feature importances.

### CODE REVIEW

**Comment by jduerholt** [comment]
@bertiqwerty: this is now ready for review. Looking forward to your comments ;)

File: bofire/models/model.py [hunk 1]
```diff
@@ -89,18 +89,43 @@ def _fit(self, X: pd.DataFrame, Y: pd.DataFrame):
         pass
 
     def cross_validate(
-        self, experiments: pd.DataFrame, folds: int = -1
-    ) -> Tuple[CvResults, CvResults]:
+        self,
+        experiments: pd.DataFrame,
+        folds: int = -1,
+        include_X: bool = False,
+        hooks: Dict[
```
Comment (bertiqwerty) [inline bofire/models/model.py:96]: 1. This is a complicated type definition. Maybe create custom data-class `Hook`?
2. Can the hook-name be an arbitrary string?

**Review by bertiqwerty — APPROVED** [review]
Thank, Johannes! I left some comments.

File: bofire/models/model.py (hunk 1 above)
Comment (jduerholt) [reply bofire/models/model.py:96]: 1. I was not exactly sure what you meat here, we can change it later ;)
2. Yes, string is arbitrary and is the name for the hook.

File: bofire/models/feature_importance.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        X = X_test
+        y = y_test
+    else:
+        X = X_train
+        y = y_train
+    return permutation_importance(model=model, X=X, y=y, n_repeats=n_repeats, seed=seed)
+
+
+def combine_permutation_importances(
+    importances: List[Dict[str, pd.DataFrame]],
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply bofire/models/feature_importance.py]: Changed it.

File: bofire/models/feature_importance.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    Args:
+        importances (List[Dict[str, pd.DataFrame]]): List of permutation importance dictionaries, one per fold.
+        metric (RegressionMetricsEnum, optional): Metric for which the data should be combined.
+            Defaults to RegressionMetricsEnum.R2
+
+    Returns:
+        pd.DataFrame: Dataframe holding the mean permutation importance per fold and feature. Can be further processed by
+            `describe`.
+    """
+    feature_keys = list(importances[0]["MAE"].columns)
```
[the remark this answers is not in the document]
Comment (jduerholt) [reply bofire/models/feature_importance.py]: Removed it ;)

**Pull request merged**